Use std::minmax() where appropriate

iidmsa opened the pull request:
#### 6b6ebe61f3cf65f499a3709af58a7643e31ba204
<pre>
Use std::minmax() where appropriate
<a href="https://bugs.webkit.org/show_bug.cgi?id=302315">https://bugs.webkit.org/show_bug.cgi?id=302315</a>
<a href="https://rdar.apple.com/164458330">rdar://164458330</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/WTF/wtf/EnumTraits.h:
(WTF::EnumValues::std::minmax):
(WTF::EnumValues::std::max): Deleted.
(WTF::EnumValues::std::min): Deleted.

Canonical link: <a href="https://commits.webkit.org/302856@main">https://commits.webkit.org/302856@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d8cccdcb25f4538a5403e827e0c8da8b8c6e124

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130412 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2683 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41366 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137830 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82002 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2688 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99358 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133359 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116782 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80057 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1891 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34912 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81089 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122415 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110443 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35417 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140307 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128865 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2268 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107875 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2517 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113124 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107778 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27439 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1925 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31567 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55432 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2543 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65931 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161879 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2361 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40368 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2564 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -382,13 +382,13 @@ class IntRange {
-            std::max(
-                std::max(m_min * other.m_min, m_min * other.m_max),
-                std::max(m_max * other.m_min, m_max * other.m_max)));
+        auto [min, max] = std::minmax({
+            m_min * other.m_min,
+            m_min * other.m_max,
+            m_max * other.m_min,
+            m_max * other.m_max
+        });
+        return IntRange(min, max);
```
This is nice, clearer than the old code and less repetitive. I am guessing it will not be any more efficient, because C++ compilers are pretty damned good at optimizing these kind of integer expressions, but can’t imagine it will be less efficient either.

iidmsa replied:
Interesting. Good to know. Thanks for the review.

darinadler commented on Source/WTF/wtf/EnumTraits.h:
```diff
@@ -70,8 +70,11 @@ template<typename E, E...> struct EnumValues;
 struct EnumValues {
-    static constexpr E max = std::max({ values... });
-    static constexpr E min = std::min({ values... });
+private:
+    static constexpr auto boundsImpl = std::minmax({ values... });
+public:
+    static constexpr E min = boundsImpl.first;
+    static constexpr E max = boundsImpl.second;
```
This isn’t obviously better. It’s OK, I guess.

iidmsa committed:
Use std::minmax() where appropriate
https://bugs.webkit.org/show_bug.cgi?id=302315
rdar://164458330

Reviewed by Darin Adler.

* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/WTF/wtf/EnumTraits.h:
(WTF::EnumValues::std::minmax):
(WTF::EnumValues::std::max): Deleted.
(WTF::EnumValues::std::min): Deleted.

Canonical link: https://commits.webkit.org/302856@main
